cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/store/profile/ProfileStore.java

```diff
@@ -0,0 +1,188 @@
+   *
+   * @param profileId the id of the profile to look up
+   * @return the profile information about the given profile
+   * @throws IOException if there was an IO error looking up the profile
+   * @throws NotFoundException if the profile is not found
+   */
+  public Profile getProfile(ProfileId profileId) throws IOException, NotFoundException {
+    return Transactionals.execute(transactional, context -> {
+      Profile profile = getMDS(context).get(getRowKey(profileId), Profile.class);
+      if (profile == null) {
+        throw new NotFoundException(profileId);
+      }
+      return profile;
+    }, IOException.class, NotFoundException.class);
+  }
+
+  /**
+   * Get the profile information in the given namespace
+   *
+   * @param namespaceId the id of the profile to look up
+   * @return the list of profiles which is in this namespace
+   * @throws IOException if there was an IO error looking up the profile
+   */
+  public List<Profile> getProfiles(NamespaceId namespaceId) throws IOException {
+    return Transactionals.execute(transactional, context -> {
+      List<Profile> profiles = getMDS(context).list(getRowKey(namespaceId), Profile.class);
+      return Collections.unmodifiableList(profiles);
+    }, IOException.class);
+  }
+
+  /**
+   * Add the profile to the profile store.
+   *
+   * @param profileId the id of the profile to add
+   * @param profile the information of the profile
+   * @throws IOException if there was an IO error adding the profile
+   * @throws AlreadyExistsException if the profile already exists
+   */
+  public void add(ProfileId profileId, Profile profile) throws IOException, AlreadyExistsException {
+    Transactionals.execute(transactional, context -> {
+      MetadataStoreDataset mds = getMDS(context);
+      // make sure that a profile doesn't exist
+      MDSKey rowKey = getRowKey(profileId);
+      Profile value = mds.get(rowKey, Profile.class);
+      if (value != null) {
+        throw new AlreadyExistsException(profileId,
+                                         String.format("Profile '%s' already exists.",
+                                                       profile.getName()));
+      }
+      mds.write(rowKey, profile);
+    }, IOException.class, AlreadyExistsException.class);
+  }
+
+  /**
+   * Deletes the profile from the profile store
+   *
+   * @param profileId the id of the profile to delete
+   * @throws IOException if there was an IO error deleting the profile
+   * @throws NotFoundException if the profile is not found
+   */
+  public void delete(ProfileId profileId) throws IOException, NotFoundException {
+    Transactionals.execute(transactional, context -> {
+      MetadataStoreDataset mds = getMDS(context);
+      MDSKey rowKey = getRowKey(profileId);
+      Profile value = getMDS(context).get(rowKey, Profile.class);
+      if (value == null) {
+        throw new NotFoundException(profileId);
+      }
+      mds.delete(getRowKey(profileId));
```

Comment:
Deletion of a profile requires the profile is in "disable" state first, right @albertshau ?

Reply:
oh yes, sorry I completely forgot about this part.